### DISCUSSION TITLE
Surrogate modeling - Keras

### DISCUSSION THREAD

**Discussion by juanbaquero4** [IDAES/idaes-pse#1480] [Q&A]
Hi everybody! I need your help!

I am trying to use Keras for surrogate modeling through IDAES library, however when I import an error message pops out as follows:

```
ImportError: Traceback (most recent call last):
  File "C:\Users\57313\miniconda3\envs\my-idaes-env\lib\site-packages\tensorflow\python\pywrap_tensorflow.py", line 70, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: DLL load failed while importing _pywrap_tensorflow_internal: Error en una rutina de inicialización de biblioteca de vínculos dinámicos (DLL).
```

I have done the following trying to overcome this problem:

1. I have installed the las version of IDAES, now I am using idaes-pse-2.5.0
2. I tried installing Tensorflow running the following codes through Jupyter Notebook, with no success:
```
pip install --upgrade pip
!pip install tensorflow
```
3. I tried installing Tensorflow running the following codes through conda, with the same previous code, with no success.

The code I am trying to run is the same as the example:

```
from idaes.core.surrogate.sampling.data_utils import split_training_validation
from idaes.core.surrogate.sampling.scaling import OffsetScaler
from idaes.core.surrogate.keras_surrogate import (
    KerasSurrogate,
    save_keras_json_hd5,
    load_keras_json_hd5,
)
from idaes.core.surrogate.plotting.sm_plotter import (
    surrogate_scatter2D,
    surrogate_parity,
    surrogate_residual,
)
from idaes.core.surrogate.surrogate_block import SurrogateBlock
from idaes.core import FlowsheetBlock
```

I immensely appreciate your help to overcome this error.

**Comment by andrewlee94** [comment]
@juanbaquero4 First, double check what version of `tensorflow` and IDAES you are using as we have recently had compatibility issues with newer version of `tensorflow` - notable IDAES v2.5 only supports `tensorflow < 2.16.1` (note that the soon to be released IDAES v2.6 should address the `tensorflow` version limitations). Given that you used `pip install tensorflow` I suspect this is going to be the cause of your issue.

If that does not work, could you send us the full traceback for the error? The part you have showed is an internal error to tensorflow which this team is unlikely to be able to help with. Also

**Reply by juanbaquero4** [reply]
Hi Andrew, thank you for your answer. I am double checking what you mentioned, hence, I installed again IDAES package on a new file and later I upgraded the version using `pip install --upgrade idaes-pse`, checking the IDAES version using `idaes --version` which is `idaes, version 2.5.0`.

I ran the following code from the exercise, which pops out this error:

```
import os
import numpy as np
import pandas as pd
import random as rn
import tensorflow as tf
```
Error message:

```
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[30], line 5
      3 import pandas as pd
      4 import random as rn
----> 5 import tensorflow as tf

ModuleNotFoundError: No module named 'tensorflow'
```

That was the reason I tried to install TensorFlow via PIP. But now, I will be waiting your suggestion of the next step to take to install TensorFlow

**Reply by andrewlee94** [reply]
@juanbaquero4 That error message simply indicates that you have not installed `tensorflow` in your current environment. The issue you have with the previous error message is because you used `pip install tensorflow` which would have installed the latest version of `tensorflow`, which is not compatible with IDAES v2.5.

Try doing `pip install tensorflow==2.16.1` instead - this will get a earlier version of `tensorflow` which is compatible with IDAES.

**Reply by juanbaquero4** [reply]
Thank you @andrewlee94 , after installing `tensorflow==2.16.1` I ran again `import tensorflow as tf` which arises the following error:

```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[1], line 5
      3 import pandas as pd
      4 import random as rn
----> 5 import tensorflow as tf

File ~\miniconda3\envs\idaes-2.5.0\lib\site-packages\tensorflow\__init__.py:42
     40 # Make sure code inside the TensorFlow codebase can use tf2.enabled() at import.
     41 _os.environ["TF2_BEHAVIOR"] = "1"
---> 42 from tensorflow.python import tf2 as _tf2
     43 _tf2.enable()
     45 from tensorflow._api.v2 import __internal__

File ~\miniconda3\envs\idaes-2.5.0\lib\site-packages\tensorflow\python\tf2.py:21
      1 # Copyright 2018 The TensorFlow Authors. All Rights Reserved.
      2 #
      3 # Licensed under the Apache License, Version 2.0 (the "License");
   (...)
     13 # limitations under the License.
     14 # ==============================================================================
     15 """Tools to help with the TensorFlow 2.0 transition.
     16 
     17 This module is meant for TensorFlow internal implementation, not for users of
     18 the TensorFlow library. For that see tf.compat instead.
     19 """
---> 21 from tensorflow.python.platform import _pywrap_tf2
     22 from tensorflow.python.util.tf_export import tf_export
     25 def enable():
     26   # Enables v2 behaviors.

ImportError: DLL load failed while importing _pywrap_tf2: Error en una rutina de inicialización de biblioteca de vínculos dinámicos (DLL).
```

**Reply by andrewlee94** [reply]
That is an internal `tensorflow` error, so it is not something we can really help with. It appears that there is a missing dll file if I am reading the message correctly, but it does not say what that might be. You will probably need to go to the `tensorflow` boards for help with that.

**Reply by juanbaquero4** [reply]
Thank you again @andrewlee94. Just checking what would be the potential issue:
1. I create another environment and just installed TensorFlow with other libraries as numpy, pandas and others... and TensorFlow worked perfectly (I am not using GPU)
2.  Taking that into account, I reckon the conflict is with a library installed in IDAES package, in that case I tried to change the version of some libraries, but it wasn't possible from me to find any incompatibility. I am still working on it...

Meanwhile, I will be working with the other environment to generate the models ...
